src/push.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as http from 'http';
+import * as validUrl from 'valid-url';
+
+/**
+ * JSON format for a multi-file push manifest.
+ */
+export interface PushManifestData {
+  [source: string]: {[target: string]: {type: string; weight?: number;}}
+}
+
+/**
+ * Maps from an HTTP request path to the set of additional resources that
+ * should be pre-emptively pushed to the client via HTTP/2 server push.
+ */
+export class PushManifest {
+  private mapping = new Map<string, Map<string, {type: string}>>();
+
+  /**
+   * Create a new `PushManifest` from a JSON object which is expected to match
+   * the multi-file variant of the format described at
+   * https://github.com/GoogleChrome/http2-push-manifest.
+   *
+   * Throws an exception if the given object does not match the manifest
+   * format, if a resource is not a valid URI path, or if `type` is not one of
+   * the valid request destinations
+   * (https://fetch.spec.whatwg.org/#concept-request-destination).
+   *
+   * Note that this class does not validate that resources exist on disk, since
+   * we can't assume if or how the server maps resources to disk.
+   */
+  constructor(manifest: PushManifestData) {
+    for (const source of Object.keys(manifest)) {
+      validatePath(source);
+      const targets = new Map();
+      for (const target of Object.keys(manifest[source])) {
+        validatePath(target);
+        const t = manifest[source][target].type || '';
+        if (!requestDestinations.has(t)) {
```

Comment:
In the next version of typescript they're adding string enums. Until then the most elegant workaround I've got is to list the enum members as exported `const`s in one file, then define the enum value and type in another file. e.g.

```typescript
// push-type-impl.ts
export const document = 'document';
export const script = 'script';
```

```typescript
// push-type.ts
import * as PushTypeValues from './push-type-impl.ts';
export const PushType = PushTypeValues;
export type PushType = (typeof PushTypeValues)[keyof typeof PushTypeValues];
```

Then you can use PushType in that module as an enum.